src/binaryen-shell.cpp

```diff
@@ -53,16 +53,67 @@ struct ExitException {
 //
 
 struct ShellExternalInterface : ModuleInstance::ExternalInterface {
-  char *memory;
+  // The underlying memory can be accessed through unaligned pointers which
+  // isn't well-behaved in C++. WebAssembly nonetheless expects it to behave
+  // properly. Avoid emitting unaligned load/store by checking for alignment
+  // explicitly, and performing memcpy if unaligned.
+  class Memory {
+    // Use char because it doesn't run afoul of aliasing rules.
+    std::vector<char> memory;
+    template <typename T>
+    bool aligned(size_t address) {
+      static_assert(!(alignof(T) & (alignof(T) - 1)), "must be a power of 2");
+      return 0 == (address & (alignof(T) - 1));
+    }
+    Memory(Memory&) = delete;
+    Memory& operator=(const Memory&) = delete;
+
+   public:
+    Memory() {}
+    void resize(size_t newSize) {
+      // Allocate at least this size to get proper alignment.
+      // The code is optimistic this will work until wg21.link/p0035r0 happens.
+      const size_t minSize = 1 << 12;
```

Comment:
The `vector`'s memory isn't guaranteed to be aligned either. A `vector<double>` would be but then you can't correct cast from it to the other types (only `char` works). We could allocate `sizeof(double)` extra and realign, but that breaks down when you `resize` causing a `realloc` which is now unaligned. All corner cases really because after a certain size the allocator plays nice anyways.